### PR TITLE
Snapshot removal and storage cleanup logs

### PR DESCRIPTION
**Pull request by hsato03**
### Description

The snapshot removal and storage cleanup logs have few information about these processes, which makes troubleshooting difficult.

This PR intends to add new log messages and rewrite the old ones to make troubleshooting easier.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [X] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

I decreased the value of the `storage.cleanup.interval` configuration and deleted a snapshot. Then I checked the management server logs.

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Review by harikrishna-patnala — APPROVED** [review]
Code LGTM

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 1]
```diff
@@ -1289,37 +1289,52 @@ public void cleanupStorage(boolean recurring) {
...
-                                s_logger.warn("Problem cleaning up primary storage pool " + pool, e);
+                                s_logger.error(String.format("Failed to clean up primary storage pool [%s] due to: [%s].", pool, e.getMessage()), e);
                             }
                         }
                     }
 
                     //destroy snapshots in destroying state in snapshot_store_ref
                     List<SnapshotDataStoreVO> ssSnapshots = _snapshotStoreDao.listByState(ObjectInDataStoreStateMachine.State.Destroying);
                     for (SnapshotDataStoreVO ssSnapshotVO : ssSnapshots) {
+                        SnapshotVO snapshot = _snapshotDao.findById(ssSnapshotVO.getSnapshotId());
```
Comment (vishesh92) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1319]: Use `getSnapshotVO` on snapshot info object to get the snapshot & it's uuid.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 2]
```diff
@@ -1289,37 +1289,52 @@ public void cleanupStorage(boolean recurring) {
...
+                            continue;
+                        }
+
+                        String snapshotUuid = snapshot.getUuid();
                         try {
-                            _snapshotService.deleteSnapshot(snapshotFactory.getSnapshot(ssSnapshotVO.getSnapshotId(), DataStoreRole.Image));
+                            s_logger.debug(String.format("Verifying if snapshot [%s] is in destroying state in any image data store.", snapshotUuid));
+                            SnapshotInfo snapshotInfo = snapshotFactory.getSnapshot(snapshot.getId(), DataStoreRole.Image);
+
+                            if (snapshotInfo != null) {
```
Comment (vishesh92) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1330]: If snapshot doesn't exist, snapshotInfo will also be null.

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java [hunk 3]
```diff
@@ -364,7 +364,7 @@ protected Void deleteSnapshotCallback(AsyncCallbackDispatcher<SnapshotServiceImp
         SnapshotResult res = null;
         try {
             if (result.isFailed()) {
-                s_logger.debug("delete snapshot failed" + result.getResult());
+                s_logger.debug(String.format("Failed to delete snapshot [%s] due to: [%s].", snapshot.getUuid(), result.getResult()));
```
[the remark this answers is not in the document]
Comment (GutoVeronezi) [reply engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java:367]: Actually, it is not a consensus.

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java [hunk 4]
```diff
@@ -373,7 +373,7 @@ protected Void deleteSnapshotCallback(AsyncCallbackDispatcher<SnapshotServiceImp
                 res = new SnapshotResult(context.snapshot, null);
             }
         } catch (Exception e) {
-            s_logger.debug("Failed to in deleteSnapshotCallback", e);
+            s_logger.error("An exception occurred while processing an event in delete snapshot callback.", e);
```
Comment (DaanHoogland) [inline engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java:376]:
```suggestion
            s_logger.error("An exception occurred while processing an event in delete snapshot callback.");
            s_logger.debug("Exception while processing an event in delete snapshot callback.", e);
```
stacktraces should not go in error level logging

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java [hunk 5]
```diff
@@ -418,15 +418,13 @@ public boolean deleteSnapshot(SnapshotInfo snapInfo) {
...
                 throw new CloudRuntimeException(result.getResult());
             }
+            s_logger.debug(String.format("Successfully deleted snapshot [%s] with ID [%s].", snapInfo.getName(), snapInfo.getUuid()));
             return true;
-        } catch (InterruptedException e) {
-            s_logger.debug("delete snapshot is failed: " + e.toString());
-        } catch (ExecutionException e) {
-            s_logger.debug("delete snapshot is failed: " + e.toString());
+        } catch (InterruptedException | ExecutionException e) {
+            s_logger.error(String.format("Failed to delete snapshot [%s] due to: [%s].", snapInfo.getUuid(), e.getMessage()), e);
```
Comment (DaanHoogland) [inline engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java:424]:
```suggestion
            s_logger.error(String.format("Failed to delete snapshot [%s] due to: [%s].", snapInfo.getUuid(), e.getMessage()));
            s_logger.error(String.format("Failed to delete snapshot [%s].", snapInfo.getUuid()), e);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 6]
```diff
@@ -1289,37 +1289,52 @@ public void cleanupStorage(boolean recurring) {
...
+                                        s_logger.debug(String.format("Storage pool garbage collector has marked template [%s] on pool [%s] " +
+                                                "for garbage collection.", templatePoolVO.getTemplateId(), templatePoolVO.getPoolId()));
                                         continue;
                                     }
 
                                     _tmpltMgr.evictTemplateFromStoragePool(templatePoolVO);
                                 }
                             } catch (Exception e) {
-                                s_logger.warn("Problem cleaning up primary storage pool " + pool, e);
+                                s_logger.error(String.format("Failed to clean up primary storage pool [%s] due to: [%s].", pool, e.getMessage()), e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1311]:
```suggestion
                                s_logger.error(String.format("Failed to clean up primary storage pool [%s] due to: [%s].", pool, e.getMessage()));
                                s_logger.error(String.format("Failed to clean up primary storage pool [%s].", pool), e);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 7]
```diff
@@ -1289,37 +1289,52 @@ public void cleanupStorage(boolean recurring) {
...
+
+                            if (snapshotInfo != null) {
+                                s_logger.debug(String.format("Snapshot [%s] in destroying state found in image data store [%s]; therefore, it will be destroyed.", snapshotUuid, snapshotInfo.getDataStore().getUuid()));
+                                _snapshotService.deleteSnapshot(snapshotInfo);
+                            } else {
+                                s_logger.debug(String.format("Did not find snapshot [%s] in destroying state in any image data store.", snapshotUuid));
+                            }
                         } catch (Exception e) {
-                            s_logger.debug("Failed to delete snapshot: " + ssSnapshotVO.getId() + " from storage");
+                            s_logger.error(String.format("Failed to delete snapshot [%s] from storage due to: [%s].", snapshotUuid, e.getMessage()), e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1337]:
```suggestion
                            s_logger.error(String.format("Failed to delete snapshot [%s] from storage due to: [%s].", snapshotUuid, e.getMessage()));
                            s_logger.debug(String.format("Failed to delete snapshot [%s] from storage.", snapshotUuid), e);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 8]
```diff
@@ -1343,7 +1358,7 @@ public void cleanupStorage(boolean recurring) {
                             // system, but not necessary.
                             handleManagedStorage(vol);
                         } catch (Exception e) {
-                            s_logger.warn("Unable to destroy host-side clustered file system " + vol.getUuid(), e);
+                            s_logger.error(String.format("Unable to destroy host-side clustered file system [%s] due to: [%s].", vol.getUuid(), e.getMessage()), e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1361]:
```suggestion
                            s_logger.error(String.format("Unable to destroy host-side clustered file system [%s] due to: [%s].", vol.getUuid(), e.getMessage()));
                            s_logger.error(String.format("Unable to destroy host-side clustered file system [%s].", vol.getUuid()), e);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 9]
```diff
@@ -1369,7 +1384,7 @@ public void cleanupStorage(boolean recurring) {
                             }
                             _snapshotDao.expunge(snapshotVO.getId());
                         } catch (Exception e) {
-                            s_logger.warn("Unable to destroy snapshot " + snapshotVO.getUuid(), e);
+                            s_logger.error(String.format("Unable to destroy snapshot [%s] due to: [%s].", snapshotVO.getUuid(), e.getMessage()), e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1387]:
```suggestion
                            s_logger.error(String.format("Unable to destroy snapshot [%s] due to: [%s].", snapshotVO.getUuid(), e.getMessage()));
                            s_logger.debug(String.format("Unable to destroy snapshot [%s].", snapshotVO.getUuid()), e);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 10]
```diff
@@ -1397,17 +1412,17 @@ public void cleanupStorage(boolean recurring) {
...
                                         if (!result.isSuccess()) {
-                                            s_logger.warn("Failed to expunge volume " + volume.getUuid() + " from the image store " + dataStore.getName() + " due to: " + result.getResult());
+                                            s_logger.warn(String.format("Failed to expunge volume [%s] from the image store [%s] due to: [%s].", volume.getUuid(), dataStore.getName(), result.getResult()));
                                         }
                                     }
                                 }
                             }
                         } catch (Throwable th) {
-                            s_logger.warn("Unable to destroy uploaded volume " + volume.getUuid() + ". Error details: " + th.getMessage());
+                            s_logger.error(String.format("Unable to destroy uploaded volume [%s] due to: [%s].", volume.getUuid(), th.getMessage()), th);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1425]:
```suggestion
                            s_logger.error(String.format("Unable to destroy uploaded volume [%s] due to: [%s].", volume.getUuid(), th.getMessage()));
                            s_logger.error(String.format("Unable to destroy uploaded volume [%s].", volume.getUuid()), th);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 11]
```diff
@@ -1456,7 +1471,7 @@ public void cleanupStorage(boolean recurring) {
                                 }
                             }
                         } catch (Throwable th) {
-                            s_logger.warn("Unable to destroy uploaded template " + template.getUuid() + ". Error details: " + th.getMessage());
+                            s_logger.error(String.format("Unable to destroy uploaded template [%s] due to: [%s].", template.getUuid(), th.getMessage()), th);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1474]:
```suggestion
                            s_logger.error(String.format("Unable to destroy uploaded template [%s] due to: [%s].", template.getUuid(), th.getMessage()));
                            s_logger.debug(String.format("Unable to destroy uploaded template [%s].", template.getUuid()), th);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java (hunk 2 above)
Comment (hsato03) [reply server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1330]: You are right. But, snapshotInfo is getting only  snapshots from Image data store role. If a snapshot is in destroying state in a Primary data store role and the same snapshot in the Image data store role has already been cleaned up, a NPE will be thrown.

Screenshot from 2023-08-09 16-28-32

Snapshots from ID 18 & 19 were cleaned up in Image data store but the snapshot still remains in destroying state in Primary data store.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java (hunk 1 above)
Comment (hsato03) [reply server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1319]: As mentioned in https://github.com/apache/cloudstack/pull/8031/files#r1353106386, snapshot info can be null.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java (hunk 2 above)
Comment (vishesh92) [reply server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1330]: As per my understanding, the below code takes care of that. We don't need to check for `snapshot` and `snapshotInfo` separately. Just a check on snapshotInfo should be fine and the code above (https://github.com/apache/cloudstack/pull/8031/files#diff-cf68e132ad7711cfa9250cda5819c09f51cba9933aa595cd82ad5a02b4edbe64R1319-R1325) can be removed.

https://github.com/apache/cloudstack/blob/b4c7705d4bc8d29f20a44c48081b80d06ab2a789/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/SnapshotDataFactoryImpl.java#L91-L111

Comment (hsato03) [reply server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1330]: The code you sugested to remove is being used to get the snapshot uuid and to log it in places that are out of `snapshotInfo` scope and in cases where `snapshotInfo` is null.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 12]
```diff
@@ -1352,10 +1370,10 @@ public void cleanupStorage(boolean recurring) {
                                 volService.ensureVolumeIsExpungeReady(vol.getId());
                                 volService.expungeVolumeAsync(volumeInfo);
                             } else {
-                                s_logger.debug("Volume " + vol.getUuid() + " is already destroyed");
+                                s_logger.debug(String.format("Volume [%s] is already destroyed.", vol.getUuid()));
                             }
                         } catch (Exception e) {
-                            s_logger.warn("Unable to destroy volume " + vol.getUuid(), e);
+                            s_logger.error(String.format("Unable to destroy volume [%s] due to: [%s].", vol.getUuid(), e.getMessage()), e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1376]:
```suggestion
                            s_logger.error(String.format("Unable to destroy volume [%s] due to: [%s].", vol.getUuid(), e.getMessage()));
                            s_logger.debug(String.format("Unable to destroy volume [%s].", vol.getUuid()), e);
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1331]: @hsato03 , if I understand correctly you do a DB query for the sole purpose of logging. That seems a stretch. An operator can do this out of bound. If at all one would want this it should be turn-offable like behind an `if (LOGGER.isDebugEnabled()) {}` guard.

**Review by DaanHoogland — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@GutoVeronezi @vishesh92 Are you allright with this PR?

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

**Review by vishesh92 — APPROVED** [review]

**Pull request merged**